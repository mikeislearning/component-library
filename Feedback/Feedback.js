import React from 'react';
import { Image, Platform, Text, TouchableOpacity, View } from 'react-native';
import { PropTypes } from 'prop-types';

import { Button, Input, InputLabel, Error, Loading, Spacer } from 'app/components';
import { images, metrics } from 'app/theme';
import { logEvent, sendSentryError } from 'app/services';
import { setFeedbackSubmitted } from 'app/utils';

import styles from './Feedback.styles';

const emojis = ['frustrated', 'confused', 'neutral', 'satisfied', 'amazed'];

export default class Feedback extends React.Component {
  static propTypes = {
    error: PropTypes.object,
    isDismissButtonsVisible: PropTypes.bool,
    isLoading: PropTypes.bool,
    isSubmitted: PropTypes.object,
    onClose: PropTypes.func,
    onSubmit: PropTypes.func.isRequired,
  };

  static defaultProps = {
    isDismissButtonsVisible: false,
    isLoading: false,
  };

  state = {
    emoji: {
      amazed: false,
      confused: false,
      frustrated: false,
      neutral: false,
      satisfied: false,
    },
    message: '',
    submitted: false,
  };

  onChangeText = (value) => {
    this.setState({
      message: value,
    });
  };

  onPressEmoji = (name) => {
    const { emoji } = this.state;
    let emojis = {};
    Object.keys(emoji).map((key) => {
      emojis = {
        ...emojis,
        // set true if it is the selected emoji and it is not already selected
        [key]: key === name && !emoji[key],
      };
    });
    logEvent('dashboard_connections_feedback_rate', { rating: name });
    this.setState({
      emoji: emojis,
    });
  };

  onClose = () => {
    this.props.onClose();
  };

  onPressDecline = () => {
    // Mark feedback as submitted so we don't ask the user to submit the form again
    logEvent('dashboard_connections_feedback_dontask');
    this.setFeedbackInStorage();
    this.onClose();
  };

  onSubmit = async () => {
    try {
      const { emoji, message } = this.state;
      const selectedEmoji = Object.keys(emoji).find((key) => emoji[key] === true);

      await this.props.onSubmit({
        message,
        opinion: selectedEmoji,
        os: Platform.OS === 'ios' ? 'iOS' : 'Android',
      });

      logEvent('dashboard_connections_feedback_submit');

      this.setState({
        submitted: true,
      });

      this.setFeedbackInStorage();
    } catch (error) {
      sendSentryError('Error submitting feedback', error);
    }
  };

  setFeedbackInStorage = () => {
    logEvent('dashboard_connections_feedback_late');
    try {
      setFeedbackSubmitted('true');
    } catch (error) {
      sendSentryError('Error setting isFeedbackSubmitted in AsyncStorage from Feedback.js', error);
    }
  };

  render() {
    const { error, isDismissButtonsVisible, isLoading, isSubmitted } = this.props;
    const { emoji, message, submitted } = this.state;
    // Display smaller emojis on smaller screens
    const emojiStyles = metrics.screenWidth < 350 ? styles.emojiSmall : {};
    // The form is considered valid whether a message provided or emoji selected
    const isValidForm =
      message.length || Object.keys(emoji).filter((key) => emoji[key] === true).length;

    return (
      <View style={styles.container}>
        {submitted && isSubmitted.id ? (
          <View style={styles.submittedContainer}>
            <Image source={images.paperPlane} style={styles.sentIcon} />
            <Text style={[styles.text, styles.title]}>Thank you</Text>
            <Text style={[styles.text, styles.title]}>for your feedback!</Text>
            {isDismissButtonsVisible && (
              <TouchableOpacity
                accessibilityLabel={'Close feedback popup'}
                activeOpacity={0.7}
                onPress={this.onClose}
              >
                <Text style={styles.closeText}>Close</Text>
              </TouchableOpacity>
            )}
          </View>
        ) : (
          <React.Fragment>
            <Text style={[styles.text, styles.title]}>Thanks for using Arrive!</Text>
            <Text style={[styles.text, styles.subTitle]}>
              How was your Arrive experience today?
            </Text>
            <View style={styles.emojiList}>
              {emojis.map((item) => {
                return (
                  <TouchableOpacity
                    accessibilityLabel={item + ' emoji'}
                    activeOpacity={0.7}
                    key={item + 'emoji'}
                    onPress={() => this.onPressEmoji(item)}
                  >
                    <Image
                      source={images[item + (emoji[item] ? 'Active' : '')]}
                      style={[styles.emoji, emojiStyles]}
                    />
                  </TouchableOpacity>
                );
              })}
            </View>

            <Spacer size="medium" />
            <InputLabel text="Write your feedback here" />
            <Input
              value={message}
              onChangeText={this.onChangeText}
              placeholder="Your message (optional)"
              multiline={true}
              textAlignVertical="top"
            />

            <Spacer size="xsmall" />
            <Text style={[styles.text, styles.description]}>
              Your feedback will only be shared with our team and will help us make improvement.
              Thank you!
            </Text>
            <Spacer size="medium" />
            <View style={styles.buttons}>
              {!isLoading ? (
                <Button disabled={!isValidForm} onPress={this.onSubmit}>
                  Submit feedback
                </Button>
              ) : (
                <Loading />
              )}

              {isDismissButtonsVisible && (
                <View style={styles.dismissButtons}>
                  <TouchableOpacity
                    accessibilityLabel="Leave feedback later"
                    activeOpacity={0.7}
                    onPress={this.onClose}
                    style={styles.dismissButton}
                  >
                    <Text style={[styles.text, styles.dismissButtonText]}>Maybe later</Text>
                  </TouchableOpacity>

                  <View style={styles.dismissButtonSeparator} />

                  <TouchableOpacity
                    accessibilityLabel="Don't ask me again about leaving feedback"
                    activeOpacity={0.7}
                    onPress={this.onPressDecline}
                    style={styles.dismissButton}
                  >
                    <Text
                      style={[styles.text, styles.dismissButtonText]}
                    >{`Don't ask me again`}</Text>
                  </TouchableOpacity>
                </View>
              )}
            </View>
            {error && (
              <React.Fragment>
                <Spacer size="small" />
                <Error error={error} />
              </React.Fragment>
            )}
          </React.Fragment>
        )}
      </View>
    );
  }
}
