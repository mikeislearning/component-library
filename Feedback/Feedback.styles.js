import { StyleSheet } from 'react-native';
import { colors, fonts } from 'app/theme';

const styles = StyleSheet.create({
  closeText: {
    color: colors.blue,
    fontFamily: fonts.family.messinaSans.regular,
    fontSize: 16,
    padding: 18,
  },
  container: {
    flex: 1,
    paddingBottom: 20,
    paddingHorizontal: 10,
    paddingTop: 15,
  },
  description: {
    color: colors.gray.medium,
    fontSize: 12,
    textAlign: 'left',
  },
  dismissButtonSeparator: {
    backgroundColor: colors.gray.medium,
    height: 3,
    marginHorizontal: 12,
    marginTop: 9,
    width: 3,
  },
  dismissButtonText: {
    color: colors.gray.medium,
    fontSize: 15,
  },
  dismissButtons: {
    flexDirection: 'row',
    justifyContent: 'center',
    paddingTop: 18,
  },
  emoji: {
    height: 40,
    width: 40,
  },
  emojiList: {
    flexDirection: 'row',
    justifyContent: 'space-evenly',
    paddingTop: 40,
  },
  emojiSmall: {
    height: 30,
    width: 30,
  },
  sentIcon: {
    height: 83,
    marginBottom: 34,
    width: 83,
  },
  subTitle: {
    marginTop: 25,
  },
  submittedContainer: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center',
    paddingVertical: 25,
  },
  text: {
    color: colors.arriveBlack,
    fontFamily: fonts.family.messinaSans.regular,
    fontSize: 16,
    textAlign: 'center',
  },
  title: {
    fontFamily: fonts.family.messinaSans.semiBold,
    fontSize: 22,
  },
});

export default styles;
