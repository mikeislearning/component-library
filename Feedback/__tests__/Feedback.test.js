import React from 'react';
import { render } from 'react-native-testing-library';

import Feedback from '../Feedback';

describe('<Feedback />', () => {
  it('renders correctly', () => {
    const feedback = render(<Feedback />);
    expect(feedback).toMatchSnapshot();
  });

  it('renders dismiss buttons', () => {
    const feedback = render(<Feedback isDismissButtonsVisible={true} onClose={() => {}} />);
    expect(feedback).toMatchSnapshot();
  });
});
