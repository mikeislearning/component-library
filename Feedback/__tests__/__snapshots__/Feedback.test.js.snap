// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Feedback /> renders correctly 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "paddingBottom": 20,
      "paddingHorizontal": 10,
      "paddingTop": 15,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#272E39",
          "fontFamily": "MessinaSans-Regular",
          "fontSize": 16,
          "textAlign": "center",
        },
        Object {
          "fontFamily": "MessinaSans-SemiBold",
          "fontSize": 22,
        },
      ]
    }
  >
    Thanks for using Arrive!
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#272E39",
          "fontFamily": "MessinaSans-Regular",
          "fontSize": 16,
          "textAlign": "center",
        },
        Object {
          "marginTop": 25,
        },
      ]
    }
  >
    How was your Arrive experience today?
  </Text>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-evenly",
        "paddingTop": 40,
      }
    }
  >
    <View
      accessibilityLabel="frustrated emoji"
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../app/images/feedback/frustrated.png",
          }
        }
        style={
          Array [
            Object {
              "height": 40,
              "width": 40,
            },
            Object {},
          ]
        }
      />
    </View>
    <View
      accessibilityLabel="confused emoji"
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../app/images/feedback/confused.png",
          }
        }
        style={
          Array [
            Object {
              "height": 40,
              "width": 40,
            },
            Object {},
          ]
        }
      />
    </View>
    <View
      accessibilityLabel="neutral emoji"
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../app/images/feedback/neutral.png",
          }
        }
        style={
          Array [
            Object {
              "height": 40,
              "width": 40,
            },
            Object {},
          ]
        }
      />
    </View>
    <View
      accessibilityLabel="satisfied emoji"
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../app/images/feedback/satisfied.png",
          }
        }
        style={
          Array [
            Object {
              "height": 40,
              "width": 40,
            },
            Object {},
          ]
        }
      />
    </View>
    <View
      accessibilityLabel="amazed emoji"
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../app/images/feedback/amazed.png",
          }
        }
        style={
          Array [
            Object {
              "height": 40,
              "width": 40,
            },
            Object {},
          ]
        }
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "justifyContent": "center",
        },
        Object {
          "marginVertical": 20,
        },
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#726E67",
          "fontFamily": "MessinaSans-Regular",
          "fontSize": 15,
          "lineHeight": 16,
          "marginBottom": 5,
        },
      ]
    }
  >
    Write your feedback here
  </Text>
  <View
    style={
      Array [
        Object {
          "marginBottom": 10,
        },
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 8,
          "paddingBottom": 10,
          "paddingHorizontal": 1,
          "paddingTop": 15,
        },
        Object {
          "backgroundColor": "#FFFFFF",
          "borderColor": "#000000",
          "borderWidth": 1,
        },
        Object {
          "backgroundColor": "#ECECEC",
          "borderColor": "#726E67",
          "fontFamily": "MessinaSans-LightItalic",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        clearButtonMode="always"
        multiline={true}
        onChangeText={[Function]}
        placeholder="Your message (optional)"
        placeholderTextColor="#726E67"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 8,
              "color": "#000000",
              "fontFamily": "MessinaSans-Regular",
              "fontSize": 16,
              "height": 50,
              "minWidth": 260,
              "paddingBottom": 0,
              "paddingHorizontal": 10,
              "paddingTop": 0,
              "textAlignVertical": "center",
            },
            Object {},
            Object {
              "height": 100,
              "lineHeight": 22,
              "paddingHorizontal": 14,
              "paddingVertical": 15,
              "textAlignVertical": "top",
            },
            Object {
              "backgroundColor": "#ECECEC",
              "borderColor": "#726E67",
              "fontFamily": "MessinaSans-LightItalic",
            },
          ]
        }
        textAlignVertical="top"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "justifyContent": "center",
        },
        undefined,
      ]
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#272E39",
          "fontFamily": "MessinaSans-Regular",
          "fontSize": 16,
          "textAlign": "center",
        },
        Object {
          "color": "#726E67",
          "fontSize": 12,
          "textAlign": "left",
        },
      ]
    }
  >
    Your feedback will only be shared with our team and will help us make improvement. Thank you!
  </Text>
  <View
    style={
      Array [
        Object {
          "justifyContent": "center",
        },
        Object {
          "marginVertical": 20,
        },
      ]
    }
  />
  <View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ECECEC",
          "borderColor": "#ECECEC",
          "borderRadius": 6,
          "borderWidth": 1,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 15,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "MessinaSans-Regular",
              "fontSize": 16,
            },
            Object {
              "color": "#FFFFFF",
              "textAlign": "center",
            },
            Object {
              "color": "#726E67",
            },
          ]
        }
      >
        Submit feedback
      </Text>
    </View>
  </View>
</View>
`;
